File: src/tools/metavalidate.cc
#include "openmeta/build_info.h"
#include "openmeta/validate.h"

#include <cstdio>
#include <cstdlib>
#include <cstring>

namespace openmeta {
namespace {

    struct ValidationSummary final {
        uint32_t files_checked = 0;
        uint32_t files_failed  = 0;
        uint32_t warnings      = 0;
        uint32_t errors        = 0;
    };


    static void usage(const char* argv0)
    {
        std::printf(
            "Usage: %s [options] <file> [file...]\n"
            "\n"
            "Validate metadata decode health and DNG/CCM coherency.\n"
            "\n"
            "Options:\n"
            "  --help                   Show this help\n"
            "  --version                Print OpenMeta build info\n"
            "  --no-build-info          Hide build info header\n"
            "  --xmp-sidecar            Also read sidecar XMP (<file>.xmp, <basename>.xmp)\n"
            "  --no-pointer-tags        Do not store pointer tags\n"
            "  --makernotes             Attempt MakerNote decode (best-effort)\n"
            "  --no-printim             Do not decode PrintIM fields\n"
            "  --no-decompress          Do not decompress payloads\n"
            "  --warnings-as-errors     Fail files that have warnings\n"
            "  --strict                 Alias for --warnings-as-errors\n"
            "  --c2pa-verify            Request draft C2PA verify scaffold evaluation\n"
            "  --c2pa-verify-backend <none|auto|native|openssl>\n"
            "                           Verification backend preference\n"
            "\n"
            "DNG/CCM validation:\n"
            "  --ccm-validation <none|dng-warnings>\n"
            "                           CCM validation mode (default: dng-warnings)\n"
            "  --ccm-no-require-dng-context\n"
            "                           Allow CCM query outside explicit DNG context\n"
            "  --ccm-no-reduction       Skip ReductionMatrix* fields\n"
            "  --max-ccm-fields N       Max CCM fields to collect (default: 128)\n"
            "  --max-ccm-values N       Max values per CCM field (default: 256)\n"
            "\n"
            "Resource limits:\n"
            "  --max-file-bytes N       Optional file mapping cap in bytes (default: 0=unlimited)\n"
            "  --max-payload-bytes N    Max reassembled/decompressed payload bytes\n"
            "  --max-payload-parts N    Max payload part count\n"
            "  --max-exif-ifds N        Max EXIF/TIFF IFD count\n"
            "  --max-exif-entries N     Max EXIF/TIFF entries per IFD\n"
            "  --max-exif-total N       Max total EXIF/TIFF entries\n"
            "  --max-exif-value-bytes N Max EXIF value bytes per tag\n"
            "  --max-xmp-input-bytes N  Max XMP packet bytes\n"
            "\n"
            "Capability legend:\n"
            "  scan   container/block discovery in file bytes\n"
            "  decode structured metadata decode into MetaStore entries\n"
            "  names  tag/key name mapping for human-readable output\n"
            "  dump   sidecar/preview export support via metadump/thumdump\n"
            "  details: docs/metadata_support.md (draft)\n",
            argv0 ? argv0 : "metavalidate");
    }


    static bool parse_u64_arg(const char* s, uint64_t* out)
    {
        if (!s || !*s || !out) {
            return false;
        }
        char* end            = nullptr;
        unsigned long long v = std::strtoull(s, &end, 10);
        if (!end || *end != '\0') {
            return false;
        }
        *out = static_cast<uint64_t>(v);
        return true;
    }


    static bool parse_u32_arg(const char* s, uint32_t* out)
    {
        if (!s || !*s || !out) {
            return false;
        }
        char* end       = nullptr;
        unsigned long v = std::strtoul(s, &end, 10);
        if (!end || *end != '\0') {
            return false;
        }
        *out = static_cast<uint32_t>(v);
        return true;
    }


    static bool parse_c2pa_verify_backend_arg(const char* s,
                                              C2paVerifyBackend* out)
    {
        if (!s || !*s || !out) {
            return false;
        }
        if (std::strcmp(s, "none") == 0) {
            *out = C2paVerifyBackend::None;
            return true;
        }
        if (std::strcmp(s, "auto") == 0) {
            *out = C2paVerifyBackend::Auto;
            return true;
        }
        if (std::strcmp(s, "native") == 0) {
            *out = C2paVerifyBackend::Native;
            return true;
        }
        if (std::strcmp(s, "openssl") == 0) {
            *out = C2paVerifyBackend::OpenSsl;
            return true;
        }
        return false;
    }


    static bool parse_ccm_validation_mode(const char* s, CcmValidationMode* out)
    {
        if (!s || !*s || !out) {
            return false;
        }
        if (std::strcmp(s, "none") == 0) {
            *out = CcmValidationMode::None;
            return true;
        }
        if (std::strcmp(s, "dng-warnings") == 0) {
            *out = CcmValidationMode::DngSpecWarnings;
            return true;
        }
        return false;
    }


    static const char* validate_status_name(ValidateStatus status) noexcept
    {
        switch (status) {
        case ValidateStatus::Ok: return "ok";
        case ValidateStatus::OpenFailed: return "open_failed";
        case ValidateStatus::TooLarge: return "too_large";
        case ValidateStatus::ReadFailed: return "read_failed";
        }
        return "unknown";
    }


    static const char* scan_status_name(ScanStatus status) noexcept
    {
        switch (status) {
        case ScanStatus::Ok: return "ok";
        case ScanStatus::OutputTruncated: return "output_truncated";
        case ScanStatus::Unsupported: return "unsupported";
        case ScanStatus::Malformed: return "malformed";
        }
        return "unknown";
    }


    static const char* payload_status_name(PayloadStatus status) noexcept
    {
        switch (status) {
        case PayloadStatus::Ok: return "ok";
        case PayloadStatus::OutputTruncated: return "output_truncated";
        case PayloadStatus::Unsupported: return "unsupported";
        case PayloadStatus::Malformed: return "malformed";
        case PayloadStatus::LimitExceeded: return "limit_exceeded";
        }
        return "unknown";
    }


    static const char* exif_status_name(ExifDecodeStatus status) noexcept
    {
        switch (status) {
        case ExifDecodeStatus::Ok: return "ok";
        case ExifDecodeStatus::OutputTruncated: return "output_truncated";
        case ExifDecodeStatus::Unsupported: return "unsupported";
        case ExifDecodeStatus::Malformed: return "malformed";
        case ExifDecodeStatus::LimitExceeded: return "limit_exceeded";
        }
        return "unknown";
    }


    static const char* exr_status_name(ExrDecodeStatus status) noexcept
    {
        switch (status) {
        case ExrDecodeStatus::Ok: return "ok";
        case ExrDecodeStatus::Unsupported: return "unsupported";
        case ExrDecodeStatus::Malformed: return "malformed";
        case ExrDecodeStatus::LimitExceeded: return "limit_exceeded";
        }
        return "unknown";
    }


    static const char* xmp_status_name(XmpDecodeStatus status) noexcept
    {
        switch (status) {
        case XmpDecodeStatus::Ok: return "ok";
        case XmpDecodeStatus::OutputTruncated: return "output_truncated";
        case XmpDecodeStatus::Unsupported: return "unsupported";
        case XmpDecodeStatus::Malformed: return "malformed";
        case XmpDecodeStatus::LimitExceeded: return "limit_exceeded";
        }
        return "unknown";
    }


    static const char* jumbf_status_name(JumbfDecodeStatus status) noexcept
    {
        switch (status) {
        case JumbfDecodeStatus::Ok: return "ok";
        case JumbfDecodeStatus::Unsupported: return "unsupported";
        case JumbfDecodeStatus::Malformed: return "malformed";
        case JumbfDecodeStatus::LimitExceeded: return "limit_exceeded";
        }
        return "unknown";
    }


    static const char* c2pa_verify_status_name(C2paVerifyStatus status) noexcept
    {
        switch (status) {
        case C2paVerifyStatus::NotRequested: return "not_requested";
        case C2paVerifyStatus::DisabledByBuild: return "disabled_by_build";
        case C2paVerifyStatus::BackendUnavailable: return "backend_unavailable";
        case C2paVerifyStatus::NoSignatures: return "no_signatures";
        case C2paVerifyStatus::InvalidSignature: return "invalid_signature";
        case C2paVerifyStatus::VerificationFailed: return "verification_failed";
        case C2paVerifyStatus::Verified: return "verified";
        case C2paVerifyStatus::NotImplemented: return "not_implemented";
        }
        return "unknown";
    }


    static const char*
    c2pa_verify_backend_name(C2paVerifyBackend backend) noexcept
    {
        switch (backend) {
        case C2paVerifyBackend::None: return "none";
        case C2paVerifyBackend::Auto: return "auto";
        case C2paVerifyBackend::Native: return "native";
        case C2paVerifyBackend::OpenSsl: return "openssl";
        }
        return "unknown";
    }


    static const char* ccm_status_name(CcmQueryStatus status) noexcept
    {
        switch (status) {
        case CcmQueryStatus::Ok: return "ok";
        case CcmQueryStatus::LimitExceeded: return "limit_exceeded";
        }
        return "unknown";
    }


    static const char* ccm_validation_mode_name(CcmValidationMode mode) noexcept
    {
        switch (mode) {
        case CcmValidationMode::None: return "none";
        case CcmValidationMode::DngSpecWarnings: return "dng_warnings";
        }
        return "unknown";
    }


    static const char*
    issue_severity_name(ValidateIssueSeverity severity) noexcept
    {
        switch (severity) {
        case ValidateIssueSeverity::Warning: return "warning";
        case ValidateIssueSeverity::Error: return "error";
        }
        return "unknown";
    }


    static void print_build_info_header()
    {
        std::string line1;
        std::string line2;
        format_build_info_lines(&line1, &line2);
        std::printf("%s\n%s\n", line1.c_str(), line2.c_str());
    }


    static void print_issue(const ValidateIssue& issue)
    {
        const char* sev = issue_severity_name(issue.severity);
        if (issue.category == "ccm") {
            std::printf("  %s[ccm] code=%s ifd=%s tag=0x%04X name=%s msg=%s\n",
                        sev, issue.code.c_str(), issue.ifd.c_str(),
                        static_cast<unsigned>(issue.tag), issue.name.c_str(),
                        issue.message.c_str());
            return;
        }

        if (!issue.message.empty() && issue.message != issue.code) {
            std::printf("  %s[%s] %s msg=%s\n", sev, issue.category.c_str(),
                        issue.code.c_str(), issue.message.c_str());
        } else {
            std::printf("  %s[%s] %s\n", sev, issue.category.c_str(),
                        issue.code.c_str());
        }
    }


    static void print_result_header(const ValidateResult& result,
                                    const ValidateOptions& options)
    {
        std::printf("size=%llu\n",
                    static_cast<unsigned long long>(result.file_size));

        if (result.status != ValidateStatus::Ok) {
            std::printf("status=%s\n", validate_status_name(result.status));
            return;
        }

        std::printf(
            "scan=%s payload=%s exif=%s xmp=%s exr=%s jumbf=%s c2pa_verify=%s c2pa_backend=%s entries=%u\n",
            scan_status_name(result.read.scan.status),
            payload_status_name(result.read.payload.status),
            exif_status_name(result.read.exif.status),
            xmp_status_name(result.read.xmp.status),
            exr_status_name(result.read.exr.status),
            jumbf_status_name(result.read.jumbf.status),
            c2pa_verify_status_name(result.read.jumbf.verify_status),
            c2pa_verify_backend_name(result.read.jumbf.verify_backend_selected),
            static_cast<unsigned>(result.entries));

        std::printf(
            "ccm=status=%s mode=%s require_dng=%s include_reduction=%s fields=%u dropped=%u issues=%u\n",
            ccm_status_name(result.ccm.status),
            ccm_validation_mode_name(options.ccm.validation_mode),
            options.ccm.require_dng_context ? "on" : "off",
            options.ccm.include_reduction_matrices ? "on" : "off",
            static_cast<unsigned>(result.ccm.fields_found),
            static_cast<unsigned>(result.ccm.fields_dropped),
            static_cast<unsigned>(result.ccm.issues_reported));
    }

}  // namespace
}  // namespace openmeta


int
main(int argc, char** argv)
{
    using namespace openmeta;

    bool show_build_info = true;
    ValidateOptions options;
    options.policy.max_file_bytes = 0;

    int first_path = 1;
    for (int i = 1; i < argc; ++i) {
        const char* arg = argv[i];
        if (!arg) {
            continue;
        }
        if (std::strcmp(arg, "--help") == 0) {
            usage(argv[0]);
            return 0;
        }
        if (std::strcmp(arg, "--version") == 0) {
            print_build_info_header();
            return 0;
        }
        if (std::strcmp(arg, "--no-build-info") == 0) {
            show_build_info = false;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--xmp-sidecar") == 0) {
            options.include_xmp_sidecar = true;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--no-pointer-tags") == 0) {
            options.include_pointer_tags = false;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--makernotes") == 0) {
            options.decode_makernote = true;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--no-printim") == 0) {
            options.decode_printim = false;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--no-decompress") == 0) {
            options.decompress = false;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--warnings-as-errors") == 0
            || std::strcmp(arg, "--strict") == 0) {
            options.warnings_as_errors = true;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--c2pa-verify") == 0) {
            options.verify_c2pa = true;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--c2pa-verify-backend") == 0 && i + 1 < argc) {
            C2paVerifyBackend backend = C2paVerifyBackend::Auto;
            if (!parse_c2pa_verify_backend_arg(argv[i + 1], &backend)) {
                std::fprintf(stderr, "invalid --c2pa-verify-backend value\n");
                return 2;
            }
            options.verify_backend = backend;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--ccm-validation") == 0 && i + 1 < argc) {
            CcmValidationMode mode = CcmValidationMode::None;
            if (!parse_ccm_validation_mode(argv[i + 1], &mode)) {
                std::fprintf(stderr, "invalid --ccm-validation value\n");
                return 2;
            }
            options.ccm.validation_mode = mode;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--ccm-no-require-dng-context") == 0) {
            options.ccm.require_dng_context = false;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--ccm-no-reduction") == 0) {
            options.ccm.include_reduction_matrices = false;
            first_path += 1;
            continue;
        }
        if (std::strcmp(arg, "--max-ccm-fields") == 0 && i + 1 < argc) {
            uint32_t v = 0;
            if (!parse_u32_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-ccm-fields value\n");
                return 2;
            }
            options.ccm.limits.max_fields = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-ccm-values") == 0 && i + 1 < argc) {
            uint32_t v = 0;
            if (!parse_u32_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-ccm-values value\n");
                return 2;
            }
            options.ccm.limits.max_values_per_field = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-file-bytes") == 0 && i + 1 < argc) {
            uint64_t v = 0;
            if (!parse_u64_arg(argv[i + 1], &v)) {
                std::fprintf(stderr, "invalid --max-file-bytes value\n");
                return 2;
            }
            options.policy.max_file_bytes = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-payload-bytes") == 0 && i + 1 < argc) {
            uint64_t v = 0;
            if (!parse_u64_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-payload-bytes value\n");
                return 2;
            }
            options.policy.payload_limits.max_output_bytes = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-payload-parts") == 0 && i + 1 < argc) {
            uint32_t v = 0;
            if (!parse_u32_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-payload-parts value\n");
                return 2;
            }
            options.policy.payload_limits.max_parts = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-exif-ifds") == 0 && i + 1 < argc) {
            uint32_t v = 0;
            if (!parse_u32_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-exif-ifds value\n");
                return 2;
            }
            options.policy.exif_limits.max_ifds = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-exif-entries") == 0 && i + 1 < argc) {
            uint32_t v = 0;
            if (!parse_u32_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-exif-entries value\n");
                return 2;
            }
            options.policy.exif_limits.max_entries_per_ifd = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-exif-total") == 0 && i + 1 < argc) {
            uint32_t v = 0;
            if (!parse_u32_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-exif-total value\n");
                return 2;
            }
            options.policy.exif_limits.max_total_entries = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-exif-value-bytes") == 0 && i + 1 < argc) {
            uint64_t v = 0;
            if (!parse_u64_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-exif-value-bytes value\n");
                return 2;
            }
            options.policy.exif_limits.max_value_bytes = v;
            i += 1;
            first_path += 2;
            continue;
        }
        if (std::strcmp(arg, "--max-xmp-input-bytes") == 0 && i + 1 < argc) {
            uint64_t v = 0;
            if (!parse_u64_arg(argv[i + 1], &v) || v == 0U) {
                std::fprintf(stderr, "invalid --max-xmp-input-bytes value\n");
                return 2;
            }
            options.policy.xmp_limits.max_input_bytes = v;
            i += 1;
            first_path += 2;
            continue;
        }
        break;
    }

    if (argc <= first_path) {
        usage(argv[0]);
        return 2;
    }

    if (show_build_info) {
        print_build_info_header();
    }

    ValidationSummary summary;
    int exit_code = 0;

    for (int argi = first_path; argi < argc; ++argi) {
        const char* path = argv[argi];
        if (!path || !*path) {
            continue;
        }

        summary.files_checked += 1U;

        const ValidateResult result = validate_file(path, options);

        std::printf("== %s\n", path);
        print_result_header(result, options);

        for (size_t i = 0; i < result.issues.size(); ++i) {
            print_issue(result.issues[i]);
        }

        if (result.failed) {
            std::printf("result=fail errors=%u warnings=%u\n",
                        static_cast<unsigned>(result.error_count),
                        static_cast<unsigned>(result.warning_count));
            summary.files_failed += 1U;
            exit_code = 1;
        } else if (result.warning_count != 0U) {
            std::printf("result=warn errors=%u warnings=%u\n",
                        static_cast<unsigned>(result.error_count),
                        static_cast<unsigned>(result.warning_count));
        } else {
            std::printf("result=ok errors=0 warnings=0\n");
        }

        summary.errors += result.error_count;
        summary.warnings += result.warning_count;
    }

    std::printf("summary files=%u failed=%u errors=%u warnings=%u strict=%s\n",
                summary.files_checked, summary.files_failed, summary.errors,
                summary.warnings, options.warnings_as_errors ? "on" : "off");

    return exit_code;
}

File: src/python/openmeta/python/metavalidate.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import sys

try:
    import openmeta
except ModuleNotFoundError:
    sys.stderr.write("error: Python module 'openmeta' was not found.\n")
    sys.stderr.write("Run this script with the same Python environment where the OpenMeta wheel is installed.\n")
    sys.stderr.write("\n")
    sys.stderr.write("Examples:\n")
    sys.stderr.write("  uv run python -m openmeta.python.metavalidate <file>\n")
    sys.stderr.write("  /path/to/venv/bin/python -m openmeta.python.metavalidate <file>\n")
    raise SystemExit(2)


def _snake(name: str) -> str:
    out = []
    for i, ch in enumerate(name):
        if ch.isupper() and i > 0:
            out.append("_")
        out.append(ch.lower())
    return "".join(out)


def _status_name(value: object) -> str:
    return _snake(getattr(value, "name", "unknown"))


def _result_name(result: dict[str, object]) -> str:
    if bool(result["failed"]):
        return "fail"
    if int(result["warning_count"]) != 0:
        return "warn"
    return "ok"


def main(argv: list[str]) -> int:
    ap = argparse.ArgumentParser(prog="metavalidate.py")
    ap.add_argument("files", nargs="+")
    ap.add_argument("--no-build-info", action="store_true", help="hide OpenMeta build info header")
    ap.add_argument("--json", action="store_true", help="print machine-readable JSON output")
    ap.add_argument("--xmp-sidecar", action="store_true", help="also read sidecar XMP (<file>.xmp, <basename>.xmp)")
    ap.add_argument("--no-pointer-tags", action="store_true", help="do not store pointer tags")
    ap.add_argument("--makernotes", action="store_true", help="attempt MakerNote decode (best-effort)")
    ap.add_argument("--no-decompress", action="store_true", help="do not decompress payloads")
    ap.add_argument("--warnings-as-errors", action="store_true", help="fail files that have warnings")
    ap.add_argument("--strict", action="store_true", help="alias for --warnings-as-errors")
    ap.add_argument("--c2pa-verify", action="store_true", help="request draft C2PA verify scaffold evaluation")
    ap.add_argument(
        "--c2pa-verify-backend",
        choices=["none", "auto", "native", "openssl"],
        default="auto",
        help="verification backend preference",
    )
    ap.add_argument(
        "--ccm-validation",
        choices=["none", "dng-warnings"],
        default="dng-warnings",
        help="CCM validation mode",
    )
    ap.add_argument(
        "--ccm-no-require-dng-context",
        action="store_true",
        help="allow CCM query outside explicit DNG context",
    )
    ap.add_argument("--ccm-no-reduction", action="store_true", help="skip ReductionMatrix* fields")
    ap.add_argument("--max-ccm-fields", type=int, default=128, help="max CCM fields to collect")
    ap.add_argument("--max-ccm-values", type=int, default=256, help="max values per CCM field")
    ap.add_argument(
        "--max-file-bytes",
        type=int,
        default=0,
        help="optional file mapping cap in bytes (0=unlimited)",
    )
    args = ap.parse_args(argv)

    warnings_as_errors = bool(args.warnings_as_errors or args.strict)

    if not args.no_build_info and not args.json:
        l1, l2 = openmeta.info_lines()
        print(l1)
        print(l2)
        print(openmeta.python_info_line())

    backend_none = getattr(openmeta.C2paVerifyBackend, "None")
    ccm_mode_none = getattr(openmeta.CcmValidationMode, "None")
    backend_map = {
        "none": backend_none,
        "auto": openmeta.C2paVerifyBackend.Auto,
        "native": openmeta.C2paVerifyBackend.Native,
        "openssl": openmeta.C2paVerifyBackend.OpenSsl,
    }
    ccm_mode_map = {
        "none": ccm_mode_none,
        "dng-warnings": openmeta.CcmValidationMode.DngSpecWarnings,
    }

    total_files = 0
    total_failed = 0
    total_errors = 0
    total_warnings = 0
    json_rows: list[dict[str, object]] = []

    for path in args.files:
        total_files += 1

        result = openmeta.validate(
            path,
            include_pointer_tags=not args.no_pointer_tags,
            decode_makernote=bool(args.makernotes),
            decode_printim=True,
            decompress=not args.no_decompress,
            include_xmp_sidecar=bool(args.xmp_sidecar),
            verify_c2pa=bool(args.c2pa_verify),
            verify_backend=backend_map[args.c2pa_verify_backend],
            warnings_as_errors=bool(warnings_as_errors),
            ccm_require_dng_context=not args.ccm_no_require_dng_context,
            ccm_include_reduction_matrices=not args.ccm_no_reduction,
            ccm_max_fields=int(args.max_ccm_fields),
            ccm_max_values_per_field=int(args.max_ccm_values),
            ccm_validation_mode=ccm_mode_map[args.ccm_validation],
            max_file_bytes=int(args.max_file_bytes),
        )

        if args.json:
            row: dict[str, object] = {
                "path": path,
                "file_size": int(result["file_size"]),
                "status": _status_name(result["status"]),
                "result": _result_name(result),
                "failed": bool(result["failed"]),
                "error_count": int(result["error_count"]),
                "warning_count": int(result["warning_count"]),
                "entries": int(result["entries"]),
                "issues": [],
            }
            if _status_name(result["status"]) == "ok":
                row.update(
                    {
                        "scan_status": _status_name(result["scan_status"]),
                        "payload_status": _status_name(result["payload_status"]),
                        "exif_status": _status_name(result["exif_status"]),
                        "xmp_status": _status_name(result["xmp_status"]),
                        "exr_status": _status_name(result["exr_status"]),
                        "jumbf_status": _status_name(result["jumbf_status"]),
                        "c2pa_verify_status": _status_name(result["jumbf_verify_status"]),
                        "c2pa_verify_backend": _status_name(result["jumbf_verify_backend"]),
                        "ccm_status": _status_name(result["ccm_status"]),
                        "ccm_mode": args.ccm_validation.replace("-", "_"),
                        "ccm_require_dng": not args.ccm_no_require_dng_context,
                        "ccm_include_reduction": not args.ccm_no_reduction,
                        "ccm_fields_found": int(result["ccm_fields_found"]),
                        "ccm_fields_dropped": int(result["ccm_fields_dropped"]),
                        "ccm_issues_reported": int(result["ccm_issues_reported"]),
                    }
                )
            issues: list[dict[str, object]] = []
            for issue in result["issues"]:
                issues.append(
                    {
                        "severity": _status_name(issue["severity"]),
                        "category": str(issue["category"]),
                        "code": str(issue["code"]),
                        "ifd": str(issue["ifd"]),
                        "name": str(issue["name"]),
                        "tag": int(issue["tag"]),
                        "message": str(issue["message"]),
                    }
                )
            row["issues"] = issues
            json_rows.append(row)
        else:
            print(f"== {path}")
            print(f"size={int(result['file_size'])}")
            print(
                "scan={scan} payload={payload} exif={exif} xmp={xmp} exr={exr} jumbf={jumbf} "
                "c2pa_verify={verify} c2pa_backend={backend} entries={entries}".format(
                    scan=_status_name(result["scan_status"]),
                    payload=_status_name(result["payload_status"]),
                    exif=_status_name(result["exif_status"]),
                    xmp=_status_name(result["xmp_status"]),
                    exr=_status_name(result["exr_status"]),
                    jumbf=_status_name(result["jumbf_status"]),
                    verify=_status_name(result["jumbf_verify_status"]),
                    backend=_status_name(result["jumbf_verify_backend"]),
                    entries=int(result["entries"]),
                )
            )
            print(
                "ccm=status={status} mode={mode} require_dng={require_dng} include_reduction={include_red} "
                "fields={fields} dropped={dropped} issues={issues}".format(
                    status=_status_name(result["ccm_status"]),
                    mode=args.ccm_validation.replace("-", "_"),
                    require_dng="off" if args.ccm_no_require_dng_context else "on",
                    include_red="off" if args.ccm_no_reduction else "on",
                    fields=int(result["ccm_fields_found"]),
                    dropped=int(result["ccm_fields_dropped"]),
                    issues=int(result["ccm_issues_reported"]),
                )
            )
            for issue in result["issues"]:
                severity = _status_name(issue["severity"])
                category = str(issue["category"])
                code = str(issue["code"])
                if category == "ccm":
                    print(
                        "  {sev}[ccm] code={code} ifd={ifd} tag=0x{tag:04X} name={name} msg={msg}".format(
                            sev=severity,
                            code=code,
                            ifd=str(issue["ifd"]),
                            tag=int(issue["tag"]),
                            name=str(issue["name"]),
                            msg=str(issue["message"]),
                        )
                    )
                else:
                    print(f"  {severity}[{category}] {code}")

        err_count = int(result["error_count"])
        warn_count = int(result["warning_count"])
        fail = bool(result["failed"])
        if fail:
            total_failed += 1
            if not args.json:
                print(f"result=fail errors={err_count} warnings={warn_count}")
        elif warn_count != 0:
            if not args.json:
                print(f"result=warn errors={err_count} warnings={warn_count}")
        elif not args.json:
            print("result=ok errors=0 warnings=0")

        total_errors += err_count
        total_warnings += warn_count

    if args.json:
        payload: dict[str, object] = {
            "strict": bool(warnings_as_errors),
            "files": json_rows,
            "summary": {
                "files": total_files,
                "failed": total_failed,
                "errors": total_errors,
                "warnings": total_warnings,
                "strict": bool(warnings_as_errors),
            },
        }
        if not args.no_build_info:
            line1, line2 = openmeta.info_lines()
            payload["build_info"] = {
                "line1": line1,
                "line2": line2,
                "python": openmeta.python_info_line(),
            }
        print(json.dumps(payload, ensure_ascii=True))
    else:
        print(
            "summary files={files} failed={failed} errors={errors} warnings={warnings} strict={strict}".format(
                files=total_files,
                failed=total_failed,
                errors=total_errors,
                warnings=total_warnings,
                strict="on" if warnings_as_errors else "off",
            )
        )
    return 1 if total_failed != 0 else 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
